src/Data/Nat/DivMod.agda

```diff
@@ -307,15 +308,15 @@ m∣n⇒o%n%m≡o%m m n@.(p * m) o (divides-refl p) = begin-equality
     o / pm * pm          ≤⟨ m/n*n≤m o pm ⟩
     o                    ∎
 
-m*n/m*o≡n/o : ∀ m n o .{{_ : NonZero o}} .{{_ : NonZero (m * o)}} →
+m*n/m*o≡n/o : ∀ m n o .{{_ : NonZero m}} .{{_ : NonZero o}} →
+              let instance _ = m*n≢0 m o in
```

Comment:
So I am very very wary of this change, and similar ones below. Here we are enforcing exactly the form that the proof of `NonZero (m * o)` must be, i.e. it must `m*n≢0 m o`, whereas before we could accept any proof of it. Now it may be the case that because we're only using it in irrelevant places, it doesn't matter, but we should think carefully about it. What is the _advantage_ of this change?

Reply:
Ah... interesting thought!

I had (mistakenly, perhaps) imagined that since we were precisely dealing with _irrelevant_ instances, that it would be OK. 

As to the advantage: clients of these proofs do/would not have to re-manufacture these irrelevant instances at call-sites. My _assumption_ being that the simpler, 'atomic' facts about `NonZero`ness of the arguments would be more likely to be in scope than the 'composite' ones being dealt with here?  Again, possibly a mistaken assumption!

Reply:
Hmmm... it doesn't invalidate your criticism/note of caution, but consider that we now have in `Data.Nat.Properties` all of the (pieces of the) proof that `NonZero (m * n)` iff `NonZero m` and `NonZero n`. So I suppose, following your line of thought, there might be pathological cases where we know `NonZero (m * n)` and `NonZero m` but not (yet) that `NonZero n`, except by a subsidiary appeal to the relevant 'projection' lemma... but I can't quite imagine such examples?

And since, AFAICT, *all* instances of `NonZero n` eventual reduce to the *ur*-instance `nonZero` (and hence to the discrimination of an occurrence of `suc` in the whnf of `n`)... I suppose that the typechecker would always (be able to) find such instance of `NonZero n`, 'just' by wh-normalisation...? Maybe that might prove more expensive than instance search? Can't quite see how, but that's your over-arching point: It's hard to predict how haywire instance search can get!

Reply:
I have misgivings about this as well. The whole point of instance arguments is that we let Agda figure them out (globally). Giving them locally is silly: if we're hand-constructing proofs, we may as well pass them in as arguments!

Reply:
Well... I don't know about "silly", but thanks for the feedback ;-)
As you might imagine, I take a different view: fully proof search is inherently unpredictable, and potentially very ill-behaved, so providing 'hint's to the inference, in the interests of narrowing the search space, seems a *perfectly* reasonable thing to do. Or at least, so I thought, until I read your comments...

Reply:
As I say below: take my comments about instances with a huge grain of salt.

Yes, giving Agda hints is quite sensible. I wish it was syntactically clearer that this is what was going on. (Giving an instance globally is not a hint, it is a statement of fact that gets loaded into a database of facts; giving it locally is indeed a hint, which perhaps should have a different syntax)